### PR TITLE
Feature: Add NotRoadTypes (NRT) 

### PR DESCRIPTION
**Pull request by andythenorth**
The 'one big diff' approach.

16 types only, PeterN has a patch for 64, but that needs finish, tested, argued about etc.  Can do that later.  We make progress by not waiting for perfect eh? :)  

Spec: https://wiki.openttd.org/Frosch/NotRoadTypes

Want to try it out? https://www.openttd.org/downloads/openttd-pullrequests/pr6811/latest.html has precompiled binaries for you! (last updated: 24th of March 2018)

### CODE REVIEW

**Comment by PeterN** [comment]
Argument for doing 64 types initially is it saves having to add a bunch of savegame conversion code.

**Comment by andythenorth** [comment]
Previous reviews by PeterN and TrueBrain took place in a ticket of the NRT fork (predated OpenTTD github).

https://github.com/andythenorth/NotRoadTypes/issues/22#issuecomment-379849291

**Comment by andythenorth** [comment]
Rebased by Alberth using Peter's July 2018 rebase
https://github.com/Alberth289346/OpenTTD/tree/nrt-block-based

This builds and runs fine, but it's not clear yet:

- if there are any compile warnings to attend to
- whether docs (landscape_grid.html etc) are accurate
- what state translation strings are in

With respect to 64 types (or not): this PR adds 64 types.  It's the better option.

Crowd-sourced opinion is that this won't have enough play-test time to make it into 1.9.0 by April 2019.  It has been play-tested quite a lot already, but eh.  Save it for 2.0? 🎁

**Review by nielsmh — DISMISSED** [review]
Assertion failure at tile_map.h line 91 running regression tests.

https://github.com/andythenorth/OpenTTD/blob/f5b1115039e41d7f75f91e7355003f10dd01b352/src/tile_map.h#L89-L93

**Comment by PeterN** [comment]
I have rebased to master, available here https://github.com/PeterN/OpenTTD/tree/nrt-block

New sprites have been split off to separate files to avoid future conflicts. I have not addressed the regression test failure.

**Comment by nielsmh** [comment]
The assertion failure mentioned above still happens in the regression tests.

**Comment by PeterN** [comment]
> 
> 
> The assertion failure mentioned above still happens in the regression tests.

Yes, #7142 has only just been merged into master!

**Comment by PeterN** [comment]
Two issues I've noticed during gameplay testing:

* Inconsistencies between roadtypes available and roadtypes buildable, resulting in misleading menus and button states. This may be something to do with date-introduced roadtypes.
* Not possible to set up one-way roads.

**Comment by PeterN** [comment]
One-way roads have now been fixed, this was due to the 64 road types addition. The regression failure is now resolved!

**Comment by TrueBrain** [comment]
The latest version at the time of writing can be found in binary form here:
https://www.openttd.org/downloads/openttd-pullrequests/pr6811/latest.html

I will try to keep it up-to-date where possible

**Comment by PeterN** [comment]
Need to fix the AI changelog as it's all under 1.9 still :-)

File: src/road.h [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+#include "core/smallvec_type.hpp"
+#include "newgrf.h"
+#include "economy_func.h"
+
+/** Roadtype flags. Starts with RO instead of R because R is used for rails */
+enum RoadTypeFlags {
+	ROTF_CATENARY = 0,                                     ///< Bit number for adding catenary
+	ROTF_NO_LEVEL_CROSSING,                                ///< Bit number for disabling level crossing
+	ROTF_NO_HOUSES,                                        ///< Bit number for setting this roadtype as not house friendly
+	ROTF_HIDDEN,                                           ///< Bit number for hiden from construction.
```
Comment (Gymnasiast) [inline src/road.h]: Typo in "hiden".
```suggestion
	ROTF_HIDDEN,                                           ///< Bit number for hidden from construction.
```

**Comment by andythenorth** [comment]
Merge it then :)

File: bin/ai/regression/tst_regression/main.nut [hunk 2]
```diff
@@ -535,14 +535,14 @@ function Regression::Prices()
...
 	print(" -Road-");
-	print("  ROADTYPE_ROAD,BT_ROAD:       " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_ROAD, AIRoad.BT_ROAD));
-	print("  ROADTYPE_ROAD,BT_DEPOT:      " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_ROAD, AIRoad.BT_DEPOT));
-	print("  ROADTYPE_ROAD,BT_BUS_STOP:   " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_ROAD, AIRoad.BT_BUS_STOP));
-	print("  ROADTYPE_ROAD,BT_TRUCK_STOP: " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_ROAD, AIRoad.BT_TRUCK_STOP));
-	print("  ROADTYPE_TRAM,BT_ROAD:       " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_TRAM, AIRoad.BT_ROAD));
-	print("  ROADTYPE_TRAM,BT_DEPOT:      " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_TRAM, AIRoad.BT_DEPOT));
-	print("  ROADTYPE_TRAM,BT_BUS_STOP:   " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_TRAM, AIRoad.BT_BUS_STOP));
-	print("  ROADTYPE_TRAM,BT_TRUCK_STOP: " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_TRAM, AIRoad.BT_TRUCK_STOP));
+	print("  ROADTYPE_ROAD,0,BT_ROAD:       " + AIRoad.GetBuildCost(AIRoad.ROADTYPE_ROAD, 0, AIRoad.BT_ROAD));
```
Comment (LordAro) [inline bin/ai/regression/tst_regression/main.nut]: spacing

File: src/script/api/script_road.hpp [hunk 3]
```diff
@@ -53,6 +58,14 @@ class ScriptRoad : public ScriptObject {
 		ROADTYPE_INVALID = -1,                 ///< Invalid RoadType.
 	};
 
+	/**
+	 * Subtypes of road known to the game.
+	 */
+	enum RoadSubType {
```
Comment (LordAro) [inline src/script/api/script_road.hpp]: All roadtypes have at least one subtype (0) right? Perhaps adding a constant for that would be nicer than writing "0" for the subtype for all the script functions? (I'm looking at the regression tests while writing this)

File: src/autoreplace_gui.cpp [hunk 4]
```diff
@@ -411,9 +426,19 @@ class ReplaceVehicleWindow : public Window {
...
+				this->GetWidget<NWidgetCore>(WID_RV_RAIL_ROAD_TYPE_DROPDOWN)->widget_data = sel_railtype == INVALID_RAILTYPE ? STR_REPLACE_ALL_RAILTYPE : GetRailTypeInfo(sel_railtype)->strings.replace_text;
+				break;
+
+			case VEH_ROAD:
+				/* Show the selected roadtype in the pulldown menu */
+				this->GetWidget<NWidgetCore>(WID_RV_RAIL_ROAD_TYPE_DROPDOWN)->widget_data = sel_rtid.IsValid() ? GetRoadTypeInfo(sel_rtid)->strings.replace_text : STR_REPLACE_ALL_ROADTYPE;
+				break;
+
+			default:
+				break;
```
Comment (LordAro) [inline src/autoreplace_gui.cpp]: not consistent with others, should be NOT_REACHED or not here at all

File: src/engine_type.h [hunk 5]
```diff
@@ -123,6 +125,7 @@ struct RoadVehicleInfo {
 	uint8 air_drag;          ///< Coefficient of air drag
 	byte visual_effect;      ///< Bitstuffed NewGRF visual effect data
 	byte shorten_factor;     ///< length on main map for this type is 8 - shorten_factor
+	RoadSubType roadsubtype;
```
Comment (LordAro) [inline src/engine_type.h]: doxyment

File: src/ground_vehicle.cpp [hunk 6]
```diff
@@ -28,7 +28,7 @@ void GroundVehicle<T, Type>::PowerChanged()
 	uint32 total_power = 0;
 	uint32 max_te = 0;
 	uint32 number_of_parts = 0;
-	uint16 max_track_speed = v->GetDisplayMaxSpeed();
+	uint16 max_track_speed = this->vcache.cached_max_speed; // Note: Different units for RV and Trains.
```
Comment (LordAro) [inline src/ground_vehicle.cpp]: why this change? comment doesn't really explain why...

File: src/newgrf.cpp [hunk 7]
```diff
@@ -5482,6 +5745,36 @@ static void RailTypeMapSpriteGroup(ByteReader *buf, uint8 idcount)
 	buf->ReadWord();
 }
 
+static void RoadTypeMapSpriteGroup(ByteReader *buf, uint8 idcount, RoadType basetype)
+{
+	uint8 *roadtypes = AllocaM(uint8, idcount);
```
Comment (LordAro) [inline src/newgrf.cpp:5766]: alloca make me sad :(
but fixing is for another PR, i think

File: src/newgrf.cpp [hunk 8]
```diff
@@ -5482,6 +5745,36 @@ static void RailTypeMapSpriteGroup(ByteReader *buf, uint8 idcount)
...
+	for (uint i = 0; i < idcount; i++) {
+		roadtypes[i] = _cur.grffile->roadtype_map[basetype][buf->ReadByte()];
```
Comment (LordAro) [inline src/newgrf.cpp]: no bounds check?

File: src/newgrf.cpp [hunk 9]
```diff
@@ -5884,13 +6185,20 @@ static void GraphicsNew(ByteReader *buf)
 	/* Load <num> sprites starting from <replace>, then skip <skip_num> sprites. */
 	grfmsg(2, "GraphicsNew: Replacing sprites %d to %d of %s (type 0x%02X) at SpriteID 0x%04X", offset, offset + num - 1, action5_type->name, type, replace);
 
+	if (type == 0x0D) _loaded_newgrf_features.shore = SHORE_REPLACE_ACTION_5;
```
Comment (LordAro) [inline src/newgrf.cpp:6215]: why was this moved?

File: src/newgrf.cpp [hunk 10]
```diff
@@ -9172,6 +9497,21 @@ static void ActivateOldShore()
 	}
 }
 
+/**
+ * Replocate the old tram depot sprites to the new position, if no new ones were loaded.
```
Comment (LordAro) [inline src/newgrf.cpp:9519]: replocate

File: src/newgrf_roadtype.cpp [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+ * @param tile %Tile containing the track. For track on a bridge this is the southern bridgehead.
+ * @param context Are we resolving sprites for the upper halftile, or on a bridge?
+ * @param rtsg Roadpart of interest
+ * @param param1 Extra parameter (first parameter of the callback, except roadtypes do not have callbacks).
+ * @param param2 Extra parameter (second parameter of the callback, except roadtypes do not have callbacks).
+ */
+RoadTypeResolverObject::RoadTypeResolverObject(const RoadtypeInfo *rti, TileIndex tile, TileContext context, RoadTypeSpriteGroup rtsg, uint32 param1, uint32 param2)
+	: ResolverObject(rti != NULL ? rti->grffile[rtsg] : NULL, CBID_NO_CALLBACK, param1, param2), roadtype_scope(*this, tile, context)
+{
+	this->root_spritegroup = rti != NULL ? rti->group[rtsg] : NULL;
```
Comment (LordAro) [inline src/newgrf_roadtype.cpp]: why not put this in the initialiser list above?

File: src/pathfinder/npf/npf.cpp [hunk 12]
```diff
@@ -860,7 +872,7 @@ static void NPFFollowTrack(AyStar *aystar, OpenListNode *current)
 
 	/* Information about the vehicle: TransportType (road/rail/water) and SubType (compatible rail/road types) */
 	TransportType type = user->type;
-	uint subtype = user->roadtypes;
+	uint subtype = user->roadtype;
```
Comment (LordAro) [inline src/pathfinder/npf/npf.cpp]: this doesn't appear to be a "subtype" - variable needs renaming?

File: src/road_cmd.cpp [hunk 13]
```diff
@@ -31,15 +31,151 @@
...
+
+/**
+ * Reset all road type information to its default values.
+ */
+void ResetRoadTypes()
+{
+	static const RoadtypeInfo empty_roadtype = {
+		{0,0,0,0,0,0},
+		{0,0,0,0,0,0},
+		{0,0,0,0,0,0,0,0,0,{},{},0,{},{}},
```
Comment (LordAro) [inline src/road_cmd.cpp]: spacing

File: src/road_cmd.cpp [hunk 14]
```diff
@@ -1174,6 +1359,33 @@ const byte _road_sloped_sprites[14] = {
...
+ * Get the sprite offset within a spritegroup.
+ * @param slope Slope
+ * @param bits Roadbits
+ * @return Offset for the sprite within the spritegroup.
+ */
+static uint GetRoadSpriteOffset(Slope slope, RoadBits bits)
+{
+	if (slope != SLOPE_FLAT) {
+		switch (slope) {
+			case SLOPE_NE: return 11;
```
Comment (LordAro) [inline src/road_cmd.cpp:1343]: magic numbers?

File: src/road_gui.cpp [hunk 15]
```diff
@@ -309,15 +266,20 @@ static bool RoadToolbar_CtrlChanged(Window *w)
 
 /** Road toolbar window handler. */
 struct BuildRoadToolbarWindow : Window {
-	int last_started_action; ///< Last started user action.
+	RoadTypeIdentifier roadtype_identifier; ///< Road type to build.
+	const RoadtypeInfo *rti;          ///< Informations about current road type
+	int last_started_action;    ///< Last started user action.
```
Comment (LordAro) [inline src/road_gui.cpp:273]: alignment

File: src/road_gui.cpp [hunk 16]
```diff
@@ -713,15 +758,32 @@ static Hotkey roadtoolbar_hotkeys[] = {
 	Hotkey('B', "bridge", WID_ROT_BUILD_BRIDGE),
 	Hotkey('T', "tunnel", WID_ROT_BUILD_TUNNEL),
 	Hotkey('R', "remove", WID_ROT_REMOVE),
+	Hotkey('C', "convert", WID_ROT_CONVERT_ROAD),
 	HOTKEY_LIST_END
 };
-HotkeyList BuildRoadToolbarWindow::hotkeys("roadtoolbar", roadtoolbar_hotkeys, RoadToolbarGlobalHotkeys);
+HotkeyList BuildRoadToolbarWindow::road_hotkeys("roadtoolbar", roadtoolbar_hotkeys, RoadToolbarGlobalHotkeys);
+
+static Hotkey tramtoolbar_hotkeys[] = {
```
Comment (LordAro) [inline src/road_gui.cpp:759]: isn't this just identical to roadtoolbar_hotkeys? Can it be reused instead of duplicating?

File: src/road_gui.cpp [hunk 17]
```diff
@@ -847,17 +916,54 @@ static WindowDesc _build_road_scen_desc(
...
+
+static const NWidgetPart _nested_build_tramway_scen_widgets[] = {
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_CLOSEBOX, COLOUR_DARK_GREEN),
+		NWidget(WWT_CAPTION, COLOUR_DARK_GREEN, WID_ROT_CAPTION), SetDataTip(STR_WHITE_STRING, STR_TOOLTIP_WINDOW_TITLE_DRAG_THIS),
+		NWidget(WWT_STICKYBOX, COLOUR_DARK_GREEN),
+	EndContainer(),
+	NWidget(NWID_HORIZONTAL),
+		NWidget(WWT_IMGBTN, COLOUR_DARK_GREEN, WID_ROT_ROAD_X),
+						SetFill(0, 1), SetMinimalSize(22, 22), SetDataTip(SPR_IMG_TRAMWAY_X_DIR, STR_ROAD_TOOLBAR_TOOLTIP_BUILD_TRAMWAY_SECTION),
```
Comment (LordAro) [inline src/road_gui.cpp:923]: 4 tabs seems like a lot.. is that consistent with what's done elsewhere?

File: src/road_gui.cpp [hunk 18]
```diff
@@ -1146,11 +1253,93 @@ static WindowDesc _tram_station_picker_desc(
...
 }
 
 void InitializeRoadGui()
 {
 	_road_depot_orientation = DIAGDIR_NW;
 	_road_station_picker_orientation = DIAGDIR_NW;
 }
+
+/**
+ * I really don't know why rail_gui.cpp has this too, shouldn't be included in the other one?
```
Comment (LordAro) [inline src/road_gui.cpp:1259]: uh...

File: src/saveload/afterload.cpp [hunk 19]
```diff
@@ -1045,18 +1061,18 @@ bool AfterLoadGame()
 							break;
 						case ROAD_TILE_DEPOT:    break;
 					}
-					SetRoadTypes(t, ROADTYPES_ROAD);
+					SB(_me[t].m7, 6, 2, ROADTYPES_ROAD);
```
Comment (LordAro) [inline src/saveload/afterload.cpp]: helper function for this?

File: src/script/api/script_road.cpp [hunk 20]
```diff
@@ -520,11 +565,11 @@ static bool NeighbourHasReachableRoad(::RoadTypes rts, TileIndex start_tile, Dia
 		entrance_dir = (::TileX(tile) == ::TileX(front)) ? (::TileY(tile) < ::TileY(front) ? 1 : 3) : (::TileX(tile) < ::TileX(front) ? 2 : 0);
 	}
 
-	uint p2 = station_id == ScriptStation::STATION_JOIN_ADJACENT ? 0 : 32;
+	uint p2 = (station_id != ScriptStation::STATION_JOIN_ADJACENT) << 2;
```
Comment (LordAro) [inline src/script/api/script_road.cpp]: think i'd still prefer a ternary for this

File: src/script/api/script_road.hpp [hunk 21]
```diff
@@ -71,6 +84,19 @@ class ScriptRoad : public ScriptObject {
...
+	/**
+	 * Get the name of a road sub type.
+	 * @param road_type The road type to get the name of.
+	 * @param road_sub_type The road sub type to get the name of.
+	 * @pre IsRoadTypeAvailable(road_type, road_sub_type).
+	 * @return The name the road type has.
+	 * @note Since there is no string with only the name of the road, the text which
+	 *  is shown in the dropdown where you can chose a road type is returned. This
+	 *  means that the name could be something like "Road construction" instead
+	 *  of just "Road".
```
Comment (LordAro) [inline src/script/api/script_road.hpp]: this seems like a deficiency... can this not be solved?

File: src/smallmap_gui.cpp [hunk 22]
```diff
@@ -478,6 +478,23 @@ static inline uint32 GetSmallMapRoutesPixels(TileIndex tile, TileType t)
 
 		const SmallMapColourScheme *cs = &_heightmap_schemes[_settings_client.gui.smallmap_land_colour];
 		return ApplyMask(cs->default_colour, &andor);
+	} else if (t == MP_ROAD) {
```
Comment (LordAro) [inline src/smallmap_gui.cpp]: should be transformed into a switch?

File: src/table/roadtypes.h [hunk 23]
```diff
@@ -0,0 +1,185 @@
...
+ * All the roadtype-specific information is stored here.
+ */
+
+#ifndef ROADTYPES_H
+#define ROADTYPES_H
+
+/**
+ * Global Roadtype definition
+ */
+static const RoadtypeInfo _original_roadtypes[] = {
```
Comment (LordAro) [inline src/table/roadtypes.h]: do these particularly need to be arrays? given there's one, and will only ever be one

File: src/town_cmd.cpp [hunk 24]
```diff
@@ -1174,6 +1206,41 @@ static bool GrowTownWithBridge(const Town *t, const TileIndex tile, const DiagDi
...
+ * @param t the tile where the house will be built
+ * @return true if at least one surrounding roadtype allows building houses here
+ */
+static inline bool RoadTypesAllowHouseHere(TileIndex t)
+{
+	static const TileIndexDiffC tiles[] = { {-1, -1}, {-1, 0}, {-1, 1}, {0, -1}, {0, 1}, {1, -1}, {1, 0}, {1, 1} };
+	const TileIndexDiffC *ptr;
+	bool allow = false;
+
+	for (ptr = tiles; ptr != endof(tiles); ++ptr) {
```
Comment (LordAro) [inline src/town_cmd.cpp]: ptr should be declared here?

**Review by LordAro — COMMENTED** [review]
So this is a beast of a PR. It's really hard to review in its current state, and personally I'd like to see the "Feature" commit split up further - there's some stuff about transparency, some other stuff i've forgotten because i've been reading this code for 30 mins now, and just generally lots of moving code about

Generally some doxygen comments are missing, would be good to get some of those filled out

File: bin/ai/regression/tst_regression/main.nut (hunk 2 above)
Comment (PeterN) [reply bin/ai/regression/tst_regression/main.nut]: This is autogenerated code

File: src/table/roadtypes.h (hunk 23 above)
Comment (PeterN) [reply src/table/roadtypes.h]: Good point. There were 2 at one point but I took the decision to undo that, based on our content-via-newgrf policy.

File: bin/ai/regression/tst_regression/main.nut (hunk 2 above)
Comment (LordAro) [reply bin/ai/regression/tst_regression/main.nut]: ...no it isn't? from where?

Comment (PeterN) [reply bin/ai/regression/tst_regression/main.nut]: Oh you're right, just confused me by being in bin :-)

Comment (PeterN) [reply bin/ai/regression/tst_regression/main.nut]: Oh wait, you are referring to the ,0, part. That's based on the existing spacing here and above for -Rail-.

File: src/script/api/script_road.hpp (hunk 3 above)
Comment (PeterN) [reply src/script/api/script_road.hpp]: This was modeled on the railtypes API which also has no default values.

File: src/script/api/script_road.hpp (hunk 21 above)
Comment (PeterN) [reply src/script/api/script_road.hpp]: Yeah, seems odd to not include it

File: src/smallmap_gui.cpp (hunk 22 above)
Comment (PeterN) [reply src/smallmap_gui.cpp]: Possibly. It does create a quite a bit of extra identation.

File: src/script/api/script_road.hpp (hunk 21 above)
Comment (PeterN) [reply src/script/api/script_road.hpp]: I think it's wrong, too, there is a string with just the name.

Comment (PeterN) [reply src/script/api/script_road.hpp]: Yeah, it was added later. I wonder if it's mandatory. Hmm.

**Comment by PeterN** [comment]
Only thing I see about transparency is to do with hiding catenary, which is done the same way as for elrail.

I would have liked it to be split up as well, however we had already wasted lots of time trying to disentangle the original branch, and then indeed find parts of that could be neatly separated.

File: src/newgrf_roadtype.cpp (hunk 11 above)
Comment (PeterN) [reply src/newgrf_roadtype.cpp]: Well, I can't get that working...
`error: member initializer 'root_spritegroup' does not name a non-static data member or base class`

File: src/saveload/afterload.cpp (hunk 19 above)
Comment (PeterN) [reply src/saveload/afterload.cpp]: No, this is old savegame conversion code. At this point it is converting a pre-tram game to post-tram format, then later on it will convert this pre-nrt format to post-nrt format.

File: src/script/api/script_road.hpp (hunk 3 above)
Comment (PeterN) [reply src/script/api/script_road.hpp]: RoadSubTypes are gone now!

**Comment by PeterN** [comment]
I guess I need to come clean on this, as it seems nobody knows: this is awaiting review and approval for merging.

**Comment by stormcone** [comment]
If you try to "autoreplace" a ship or an aircraft, the game crashes:
"NOT_REACHED triggered at line 460 of OpenTTD-nrt-block/src/autoreplace_gui.cpp"

**Comment by PeterN** [comment]
@stormcone Thanks, this has now been fixed.

**Review by michicc — APPROVED** [review]

**Comment by Gymnasiast** [comment]
Congratulations on getting this massive update finished! :tada:
